CMS - Multiple Choice Responses View

dandrabik opened the pull request:
## WHAT
A follow-on PR to #8061. Add an additional materialized view to power the `multiple_choice_options` endpoint.
## WHY
After testing production data, it's not possible to power the `multiple_choice_options` from just `GradedResponse` since there are not two nonoptimal-graded responses for every question.
## HOW
Abstract some of the code used for `GradedResponse` and add another materialized view migration using `scenic`.

**UPDATE**
Additionally, to have this query finish on production, I needed to scope it to responses that have a `count > 10`. I've added some logic in the controller endpoint to fallback on other responses if a question doesn't have enough responses that fit this setup. So unfortunately, the DB constraints made the code a touch more complicated.


PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  YES
Have you deployed to Staging? | YES
Self-Review: Have you done an initial self-review of the code below on Github? | Yup
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | Yes.

dandrabik committed:
Working version of multiple choice view.

dandrabik committed:
Update controller and tests.

dandrabik committed:
Revert bad commit.

dandrabik committed:
Update classname in migrations.

brendanshean commented on services/QuillCMS/app/workers/refresh_responses_view_worker.rb:
```diff
@@ -1,15 +1,21 @@
   include Sidekiq::Worker
   sidekiq_options queue: SidekiqQueue::LOW
 
   REFRESH_TIMEOUT = ENV["VIEW_STATEMENT_TIMEOUT"] || '10min'
```
Would fetch apply here:  `ENV.fetch("VIEW_STATEMENT_TIMEOUT", '10min')`

dandrabik replied:
Yup, good call, will replace.

brendanshean commented on services/QuillCMS/db/migrate/20210812215158_create_multiple_choice_responses.rb:
```diff
@@ -0,0 +1,28 @@
+  end
+
+  def down
+    drop_view :multiple_choice_responses, materialized: true
```
do the three `add_index` methods need corresponding `drop_index` in the `down` step?

dandrabik replied:
Good question! In my testing, using only `drop_view` doesn't error, but you're right, it's probably better to add the drop_indexes to ensure they are cleaned up.

brendanshean commented on services/QuillCMS/app/workers/refresh_responses_view_worker.rb:
```diff
@@ -1,15 +1,21 @@
   # Make the DB statement timeout longer while refreshing the materialized view
-  def perform
+  def perform(view)
     original_timeout = db_timeout
```
Is there ever a case where we don't want to extend the timeout on a materialized view?  In other words, should the `db_timeout` wrapper be in the concern `refresh` method (as opposed to requiring any future workers to make the call)?

I also noticed another configuration via the `establish_connection` [here](https://github.com/scenic-views/scenic/issues/201#issuecomment-434993888)

dandrabik replied:
To be honest, I went back and forth on where to keep the wrapper. I kept it in the job since we don't need the wrapper for `test` and `development` environments, it's only needed on `production` and `staging`.  The test suite has to call `refresh` relatively often, so I thought this made tests slightly less complicated. I could be convinced otherwise.

Also, I saw the `establish_connection` method, but that's for every call by the model, even queries, so I'd rather keep this scoped to refresh.

brendanshean replied:
Those are good reasons to keep it where you have it, especially the point about queries.

brendanshean commented on services/QuillCMS/spec/models/graded_response_spec.rb:
```diff
@@ -2,9 +2,9 @@
-    let!(:response_optimal) {create(:response, question_uid: '123', optimal: true)}
-    let!(:response_nonoptimal) {create(:response, question_uid: '123', optimal: false)}
-    let!(:response_ungraded) {create(:response, question_uid: '123', optimal: nil)}
+    let!(:optimal) {create(:response, question_uid: '123', optimal: true)}
```
Maybe time for factory aliases for optimal_response, non_optimal_response, etc.

dandrabik replied:
Hmm, maybe. I'm a bit on the fence here, since being explicit inline is about the same length as the abstraction (and some of the names can be confusing, i.e. distinguishing between `optimal` `false` and `nil`, which many times are treated the same). Let me think about this.

```ruby
create(:optimal_response)
create(:response, optimal: true)

brendanshean replied:
While length is about the same,  `create(:optimal_response)` prevents the accidentally specifying wrong value for `optimal`.  It also makes the `let(:optimal) { create(:optimal_response) }` easier to read.

brendanshean replied:
One other point is that it provides an opportunity to standardize language around what's being created.

dandrabik replied:
Ok, you've made several good points, I'll add the aliases. Thanks for the feedback on this.

brendanshean reviewed (approved):
Looks good to me.   Some comments you can take or leave.

dandrabik committed:
Drop indexes in down migrations.

dandrabik committed:
Use fetch.

dandrabik committed:
Add factory aliases for response types.

dandrabik committed:
Use proper remove_index method in migration.

dandrabik committed:
Add fallback and optimized query.

dandrabik commented:
I added an update after testing the queries a bit. Re-opening this for review:



> UPDATE
> Additionally, to have this query finish on production, I needed to scope it to responses that have a count > 10. I've added some logic in the controller endpoint to fallback on other responses if a question doesn't have enough responses that fit this setup. So unfortunately, the DB constraints made the code a touch more complicated.

brendanshean commented on services/QuillCMS/spec/controllers/questions_controller_spec.rb:
```diff
@@ -116,5 +118,34 @@
     end
+
+    context 'fallback responses needed' do
+      let!(:nonoptimal) {create(:graded_nonoptimal_response, question_uid: '123', count: 17)}
```
I would rename
`nonoptimal` -> `graded_nonoptimal`

brendanshean commented on services/QuillCMS/spec/controllers/questions_controller_spec.rb:
```diff
@@ -116,5 +118,34 @@
+        json = JSON.parse(response.body)
+        optimal_count = json.count {|gr| gr['optimal']}
+        nonoptimal_count = json.count {|gr| gr['optimal'] == false}
+        null_optimal_count = json.count {|gr| gr['optimal'].nil?}
```
I would rename:
`nonoptimal_count` -> `graded_nonoptimal_count`
`null_optimal_count` -> `ungraded_count`

brendanshean commented on services/QuillCMS/spec/controllers/questions_controller_spec.rb:
```diff
@@ -14,9 +14,9 @@
-      let!(:response_nonoptimal) {create(:response, question_uid: '123', optimal: false)}
-      let!(:response_ungraded) {create(:response, question_uid: '123', optimal: nil)}
+      let!(:optimal) {create(:optimal_response, question_uid: '123')}
+      let!(:nonoptimal) {create(:graded_nonoptimal_response, question_uid: '123')}
```
I would rename 
`nonoptimal` ->`graded_nonoptimal`

brendanshean commented on services/QuillCMS/spec/controllers/questions_controller_spec.rb:
```diff
@@ -46,35 +46,36 @@
-      let!(:nonoptimal1) {create(:response, question_uid: '123', optimal: false, count: 7)}
-      let!(:nonoptimal2) {create(:response, question_uid: '123', optimal: false, count: 5)}
-      let!(:nonoptimal3) {create(:response, question_uid: '123', optimal: false, count: 9)}
+      let!(:nonoptimal1) {create(:graded_nonoptimal_response, question_uid: '123', count: 17)}
```
I would rename 
`nonoptimal1` -> `graded_nonoptimal1`
`nonoptimal2` -> `graded_nonoptimal2`
`nonoptimal3` -> `graded_nonoptimal3`

brendanshean commented on services/QuillCMS/spec/controllers/questions_controller_spec.rb:
```diff
@@ -46,35 +46,36 @@
         json = JSON.parse(response.body)
         optimal_count = json.count {|gr| gr['optimal']}
-        nonoptimal_count = json.count {|gr| !gr['optimal']}
+        nonoptimal_count = json.count {|gr| gr['optimal'] == false }
```
I would rename:
`nonoptimal_count` -> `graded_nonoptimal_count` 
`null_optimal_count` -> `ungraded_count`

brendanshean commented on services/QuillCMS/spec/models/graded_response_spec.rb:
```diff
@@ -2,9 +2,9 @@
-    let!(:response_nonoptimal) {create(:response, question_uid: '123', optimal: false)}
-    let!(:response_ungraded) {create(:response, question_uid: '123', optimal: nil)}
+    let!(:optimal) {create(:optimal_response, question_uid: '123')}
+    let!(:nonoptimal) {create(:graded_nonoptimal_response, question_uid: '123')}
```
Rename:
`nonoptimal` -> `graded_non_optimal`

brendanshean commented on services/QuillCMS/spec/models/graded_response_spec.rb:
```diff
@@ -17,10 +17,12 @@
-      expect(graded_responses.first.id).to be response_optimal.id
-      expect(graded_responses.second.id).to be response_nonoptimal.id
+      expect(response_ids).to eq graded_ids
+      expect(responses.first.attributes.keys.sort).to eq nonoptimal.attributes.keys.sort
```
I'm not I understand the purpose of this expectation.

dandrabik replied:
This is confirmation that the `Response` model json and the `GradedResponse` model json have the exact same keys. If an attribute in the `Response` model changes, say we add a column via a migration, this test would fail, marking that we need to update the `GradedResponse` model, via and equivalent migration.

dandrabik replied:
Also, this may not be obvious, but the columns/attributes of the materialized view are locked after the view migration, i.e. you can't do `responses.*` in the query and expect new columns to come in when the original table is migrated. You also have to migrate the view.

https://github.com/scenic-views/scenic#faqs

brendanshean replied:
Thanks for the explanation.  I was unaware of that.   I think it's worth documenting that in its own spec:

```
it 'should have the same keys as Responses table' do
  GradedResponse.refresh
  responses = GradedResponse.where(question_uid: '123').sort_by(&:id)
  expect(responses.first.attributes.keys.sort).to eq graded_nonoptimal.attributes.keys.sort
end
```
and then add  same check for `multiple_choice_response_spec.rb` too?

brendanshean commented on services/QuillCMS/app/controllers/questions_controller.rb:
```diff
@@ -6,11 +6,37 @@ def responses
+
+  # The MultipleChoiceResponse model contains responses
+  # that have a count > MultipleChoiceResponse::MIN_COUNT
+  # All questions, like new ones, may may not have enough responses,
```
typo with "may may"

brendanshean commented on services/QuillCMS/spec/models/multiple_choice_response_spec.rb:
```diff
@@ -0,0 +1,54 @@
+    let!(:ungraded2) {create(:ungraded_response, question_uid: '123', count: 17)}
+    let!(:optimal) {create(:optimal_response, question_uid: '123', count: 9999)}
+
+
```
whitespace

brendanshean commented on services/QuillCMS/spec/models/multiple_choice_response_spec.rb:
```diff
@@ -0,0 +1,54 @@
+RSpec.describe MultipleChoiceResponse, type: :model do
+  context 'basic queries' do
+    let!(:ungraded) {create(:ungraded_response, question_uid: '123', count: 15)}
+    let!(:nonoptimal) {create(:graded_nonoptimal_response, question_uid: '123', count: 19)}
```
I would rename 
`nonoptimal` -> `graded_nonoptimal`

brendanshean commented on services/QuillCMS/spec/models/multiple_choice_response_spec.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  context 'rank ties' do
+    let!(:ungraded) {create(:ungraded_response, question_uid: '123', count: 15)}
```
Since the variables in 33-36 aren't used in the spec itself, you could throw them in a `before` clause.

brendanshean commented on services/QuillCMS/spec/models/multiple_choice_response_spec.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  context 'records with less than 10 count' do
+    let!(:nonoptimal_less_than10) {create(:graded_nonoptimal_response, question_uid: '123', count: 10)}
```
Since the variables in 46-47 aren't used in the spec itself, you could throw them in a `before` clause.

brendanshean reviewed (approved):
I think this is really close.  Great work here, this has been fun to see the development of this materialized view.

Most of the comments are on the redefining the scope [nonoptimal](https://github.com/empirical-org/Empirical-Core/pull/8078/commits/53e1f97fc1927254bcd09f274c24ab911245eb96#diff-2e42f05e854c93cd8b96d485035062fef859b6fc9bb5e38345add73874c8fd5fR10).  Since `nonoptimal` now covers two states: `false` and `nil`, renaming  `let(:nonoptimal)` to `let(:graded_nonoptimal)` would be more informative.  I'm not suggesting a similar renaming for `ungraded` to `ungraded_nonoptimal` because that would be redundant.

brendanshean commented:
@dandrabik This [PR](https://github.com/empirical-org/Empirical-Core/pull/8089) should fix those failing specs.

dandrabik committed:
Spec naming updates based on PR feedback.

dandrabik committed:
Update comment.

dandrabik commented:
> I think this is really close. Great work here, this has been fun to see the development of this materialized view.
> 
> Most of the comments are on the redefining the scope [nonoptimal](https://github.com/empirical-org/Empirical-Core/pull/8078/commits/53e1f97fc1927254bcd09f274c24ab911245eb96#diff-2e42f05e854c93cd8b96d485035062fef859b6fc9bb5e38345add73874c8fd5fR10). Since `nonoptimal` now covers two states: `false` and `nil`, renaming `let(:nonoptimal)` to `let(:graded_nonoptimal)` would be more informative. I'm not suggesting a similar renaming for `ungraded` to `ungraded_nonoptimal` because that would be redundant.

Fair points about making the naming more explicit, I've made those changes. Thanks for taking a close look at this!

dandrabik committed:
Merge branch 'develop' into feat/cms_multiple_choice
